Add more links to MSTest api docs (#44716)
dotnet/docs: docs/core/testing/unit-testing-mstest-writing-tests-attributes.md

### `TestClassAttribute`

> The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestClassAttribute> attribute marks a class that contains tests and, optionally, initialize or cleanup methods.

Reviewer: I actually prefer the original display here. It still uses an xref and the word "attribute" isn't repeated. (When you use an xref like that (with custom text), you don't actually need the angle brackets.)

> The [TestClass]](xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestClassAttribute) attribute marks a class that contains tests and, optionally, initialize or cleanup methods.

Reviewer: ```suggestion
The [TestClass](xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestClassAttribute) attribute marks a class that contains tests and, optionally, initialize or cleanup methods.
```

### `TestMethodAttribute`

The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestMethodAttribute> attribute is used inside a `TestClass` to define the actual test method to run.

The method should be an instance `public` method defined as `void`, `Task`, or `ValueTask` (starting with MSTest v3.3). It can optionally be `async` but should not be `async void`.

> The method should have zero parameters, unless it's used with <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute>, <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute> or similar attribute that provides test case data to the test method.

Reviewer: ```suggestion
The method should have zero parameters, unless it's marked with <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute>, <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute>, or a similar attribute that provides test case data to the test method.
```

> The method should have zero parameters, unless it's marked with the [DataRow](xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute) attribute, the [DynamicData])xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute) attribute, or a similar attribute that provides test case data to the test method.

Reviewer: ```suggestion
The method should have zero parameters, unless it's marked with the [DataRow](xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute) attribute, the [DynamicData](xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute) attribute, or a similar attribute that provides test case data to the test method.
```

### Class level

<xref:Microsoft.VisualStudio.TestTools.UnitTesting.ClassInitializeAttribute> is called right before your class is loaded (but after static constructor) and <xref:Microsoft.VisualStudio.TestTools.UnitTesting.ClassCleanupAttribute> is called right after your class is unloaded.

> It's possible to control the inheritance behavior: only for current class using [InheritanceBehavior.None](<xref:microsoft.visualstudio.testtools.unittesting.inheritancebehavior>) or for all derived classes using [InheritanceBehavior.BeforeEachDerivedClass](<xref:microsoft.visualstudio.testtools.unittesting.inheritancebehavior>).

Reviewer 2: ```suggestion
It's possible to control the inheritance behavior: only for current class using [InheritanceBehavior.None](<xref:Microsoft.VisualStudio.TestTools.UnitTesting.InheritanceBehavior>) or for all derived classes using [InheritanceBehavior.BeforeEachDerivedClass](<xref:Microsoft.VisualStudio.TestTools.UnitTesting.InheritanceBehavior>).
```

Reviewer: Yes, casing matters for xrefs.

> It's possible to control the inheritance behavior: only for current class using [InheritanceBehavior.None](<xref:Microsoft.VisualStudio.TestTools.UnitTesting.InheritanceBehavior>) or for all derived classes using [InheritanceBehavior.BeforeEachDerivedClass](<xref:Microsoft.VisualStudio.TestTools.UnitTesting.InheritanceBehavior>).

Reviewer: ```suggestion
It's possible to control the inheritance behavior: only for current class using <xref:Microsoft.VisualStudio.TestTools.UnitTesting.InheritanceBehavior.None?displayProperty=nameWithType>, or for all derived classes using <xref:Microsoft.VisualStudio.TestTools.UnitTesting.InheritanceBehavior.BeforeEachDerivedClass?displayProperty=nameWithType>.
```

### `STATestClassAttribute`

When applied to a test class, the [STATestClass](xref:Microsoft.VisualStudio.TestTools.UnitTesting.STATestClassAttribute) attribute indicates that all test methods (and the `[ClassInitialize]` and `[ClassCleanup]` methods) in the class should be run in a single-threaded apartment (STA). This attribute is useful when the test methods interact with COM objects that require STA.

> [!NOTE]
> This is only supported on Windows and in version 3.6 and later.

### `STATestMethodAttribute`

When applied to a test method, the [STATestMethod](xref:Microsoft.VisualStudio.TestTools.UnitTesting.STATestMethodAttribute) attribute indicates that the test method should be run in a single-threaded apartment (STA). This attribute is useful when the test method interacts with COM objects that require STA.

> [!NOTE]
> This is only supported on Windows and in version 3.6 and later.

### `ParallelizeAttribute`

> By default, MSTest runs tests in a sequential order. The assembly level attribute [Parallelize](xref:Microsoft.VisualStudio.TestTools.UnitTesting.ParallelizeAttribute) can be used to run tests in parallel. You can specify if the parallelism should be at [class level](xref:Microsoft.VisualStudio.TestTools.UnitTesting.ExecutionScope.ClassLevel?displayProperty=nameWithType) (multiple classes can be run in parallel but tests in a given class are run sequentially) or at [method level](xref:Microsoft.VisualStudio.TestTools.UnitTesting.ExecutionScope.MethodLevel?displayProperty=nameWithType).

Reviewer: ```suggestion
By default, MSTest runs tests in a sequential order. The assembly level attribute [Parallelize](xref:Microsoft.VisualStudio.TestTools.UnitTesting.ParallelizeAttribute) attribute can be used to run tests in parallel. You can specify if the parallelism should be at [class level](xref:Microsoft.VisualStudio.TestTools.UnitTesting.ExecutionScope.ClassLevel) (multiple classes can be run in parallel but tests in a given class are run sequentially) or at [method level](xref:Microsoft.VisualStudio.TestTools.UnitTesting.ExecutionScope.MethodLevel).
```

When using the timeout feature, a separate thread/task is created to run the test method. The main thread/task is responsible for monitoring the timeout and unobserving the method thread/task if the timeout is reached.

> Starting with MSTest 3.6, it is possible to specify <xref:microsoft.visualstudio.testtools.unittesting.timeoutattribute.cooperativecancellation> property on the attribute (or globally through runsettings) to enable cooperative cancellation. In this mode, the method is responsible for checking the cancellation token and aborting the test if it is signaled as you would do in a typical `async` method. This mode is more performant and allows for more precise control over the cancellation process. This mode can be applied to both async and sync methods.

Reviewer 2: ```suggestion
Starting with MSTest 3.6, it is possible to specify <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TimeoutAttribute.CooperativeCancellation*> property on the attribute (or globally through runsettings) to enable cooperative cancellation. In this mode, the method is responsible for checking the cancellation token and aborting the test if it is signaled as you would do in a typical `async` method. This mode is more performant and allows for more precise control over the cancellation process. This mode can be applied to both async and sync methods.
```

### `STATestClassAttribute`

> When applied to a test class, the <xref:microsoft.visualstudio.testtools.unittesting.statestclassattribute> attribute indicates that all test methods (and the `[ClassInitialize]` and `[ClassCleanup]` methods) in the class should be run in a single-threaded apartment (STA). This attribute is useful when the test methods interact with COM objects that require STA.

Reviewer 2: ```suggestion
When applied to a test class, the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.STATestClassAttribute> attribute indicates that all test methods (and the `[ClassInitialize]` and `[ClassCleanup]` methods) in the class should be run in a single-threaded apartment (STA). This attribute is useful when the test methods interact with COM objects that require STA.
```

> [!NOTE]
> This is only supported on Windows and in version 3.6 and later.

### `STATestMethodAttribute`

> When applied to a test method, the <xref:microsoft.visualstudio.testtools.unittesting.statestmethodattribute> attribute indicates that the test method should be run in a single-threaded apartment (STA). This attribute is useful when the test method interacts with COM objects that require STA.

Reviewer 2: ```suggestion
When applied to a test method, the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.STATestMethodAttribute> attribute indicates that the test method should be run in a single-threaded apartment (STA). This attribute is useful when the test method interacts with COM objects that require STA.
```

> [!NOTE]
> This is only supported on Windows and in version 3.6 and later.

### `ParallelizeAttribute`

By default, MSTest runs tests in a sequential order. The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.ParallelizeAttribute> attribute can be used to run tests in parallel. This is an assembly level attribute. You can specify if the parallelism should be at **class level** (multiple classes can be run in parallel but tests in a given class are run sequentially) or at **method level**.

It's also possible to specify the maximum number of threads to use for parallel execution. A value of `0` (default value) means that the number of threads is equal to the number of logical processors on the machine.

It is also possible to specify the parallelism through the [parallelization properties of the runsettings file](./unit-testing-mstest-configure.md#mstest-element).

### `DoNotParallelizeAttribute`

The <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DoNotParallelizeAttribute> attribute can be used to prevent parallel execution of tests in a given assembly. This attribute can be applied at the assembly level, class level or method level.

> [!NOTE]

> > By default, MSTest runs tests in sequential order so you only need to use this attribute if you have applied the <xref:microsoft.visualstudio.testtools.unittesting.parallelizeattribute> attribute at the assembly level.

Reviewer 2: ```suggestion
> By default, MSTest runs tests in sequential order so you only need to use this attribute if you have applied the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.ParallelizeAttribute> attribute at the assembly level.
```

### `ExpectedExceptionAttribute`

> The [ExpectedException](xref:Microsoft.VisualStudio.TestTools.UnitTesting.ExpectedExceptionAttribute) attribute defines the expected exception the test method will throw. The test will pass if the expected exception is thrown and the exception message matches the expected message.

Reviewer: ```suggestion
The [ExpectedException](xref:Microsoft.VisualStudio.TestTools.UnitTesting.ExpectedExceptionAttribute) attribute defines the exception that the test method is expected to throw. The test passes if the expected exception is thrown and the exception message matches the expected message.
```
